src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -166,6 +168,16 @@ inline TShape ReduceAxesShapeImpl(const TShape& ishape, const TShape& axis,
     }
   }
 
+  for (index_t i = 0; i < axis.ndim(); ++i) {
+    int64_t value = axis[i];
+    if (value < 0) {
+      CHECK_LT(-value-1, ishape.ndim())
+        << "Reduction axis " << value
+        << " Exceeds input dimensions " << ishape;
+      ((TShape&)axis)[i] = value + ishape.ndim();
```

Comment:
This cast is ugly. Don't do this.
Also ReduceAxesShapeImpl assumes axis is sorted. Modifying axis violates that assumption